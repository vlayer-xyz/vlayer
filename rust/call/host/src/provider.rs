use alloy_primitives::{Address, BlockNumber, Bytes, StorageKey, StorageValue, TxNumber, U256};
use auto_impl::auto_impl;
use call_engine::block_header::EvmBlockHeader;
use std::error::Error as StdError;

pub mod cache;
pub mod ethers;
pub mod factory;
mod file;
pub mod multi;
mod null;

use crate::proof::EIP1186Proof;
pub use ethers::{EthersProvider, EthersProviderError};
use ethers_core::types::BlockNumber as BlockTag;
use ethers_providers::{Http, RetryClient};
pub use file::FileProvider;

/// The Ethers client type.
pub type EthersClient = ethers_providers::Provider<RetryClient<Http>>;

/// A trait for providers that fetch data from the Ethereum blockchain.
#[auto_impl(Rc)]
pub trait BlockingProvider {
    type Error: StdError + Send + Sync + 'static;

    fn get_balance(&self, address: Address, block: BlockNumber) -> Result<U256, Self::Error>;
    fn get_block_header(
        &self,
        block: BlockTag,
    ) -> Result<Option<Box<dyn EvmBlockHeader>>, Self::Error>;
    fn get_code(&self, address: Address, block: BlockNumber) -> Result<Bytes, Self::Error>;
    fn get_proof(
        &self,
        address: Address,
        storage_keys: Vec<StorageKey>,
        block: BlockNumber,
    ) -> Result<EIP1186Proof, Self::Error>;
    fn get_storage_at(
        &self,
        address: Address,
        key: StorageKey,
        block: BlockNumber,
    ) -> Result<StorageValue, Self::Error>;
    fn get_transaction_count(
        &self,
        address: Address,
        block: BlockNumber,
    ) -> Result<TxNumber, Self::Error>;
}
