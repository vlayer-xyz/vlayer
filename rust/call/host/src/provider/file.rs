use super::{cache::CachedProvider, null::NullProvider};
use crate::provider::cache::json::JsonCache;
use std::{cell::RefCell, marker::PhantomData, path::PathBuf};

/// A provider returning responses cached in a file.
/// It panics if queries are not found in the cache.
pub(crate) type FileProvider = CachedProvider<NullProvider>;

impl FileProvider {
    /// Creates a new [FileProvider] loading the given file.
    pub(crate) fn from_file(file_path: &PathBuf) -> anyhow::Result<Self> {
        let cache = JsonCache::load(file_path)?;
        Ok(Self {
            inner: NullProvider(PhantomData),
            cache: RefCell::new(cache),
        })
    }
}
