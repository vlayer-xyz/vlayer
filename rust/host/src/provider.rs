use alloy_primitives::{
    Address, BlockNumber, Bytes, StorageKey, StorageValue, TxNumber, B256, U256,
};
use ethers::{from_ethers_bytes, from_ethers_u256};
use ethers_core::types::StorageProof as EthersStorageProof;
use serde::{Deserialize, Serialize};
use std::{error::Error as StdError, fmt::Debug};
use vlayer_engine::evm::block_header::EvmBlockHeader;

mod cache;
mod ethers;
mod file;
mod null;

pub use ethers::{EthersProvider, EthersProviderError};
use ethers_providers::{Http, RetryClient};
pub use file::{EthFileProvider, FileProvider};

/// The Ethers client type.
pub type EthersClient = ethers_providers::Provider<RetryClient<Http>>;

/// A trait for providers that fetch data from the Ethereum blockchain.
pub trait Provider {
    type Error: StdError + Send + Sync + 'static;
    type Header: EvmBlockHeader;

    fn get_block_header(&self, block: BlockNumber) -> Result<Option<Self::Header>, Self::Error>;
    fn get_transaction_count(
        &self,
        address: Address,
        block: BlockNumber,
    ) -> Result<TxNumber, Self::Error>;
    fn get_balance(&self, address: Address, block: BlockNumber) -> Result<U256, Self::Error>;
    fn get_code(&self, address: Address, block: BlockNumber) -> Result<Bytes, Self::Error>;
    fn get_storage_at(
        &self,
        address: Address,
        key: StorageKey,
        block: BlockNumber,
    ) -> Result<StorageValue, Self::Error>;
    fn get_proof(
        &self,
        address: Address,
        storage_keys: Vec<StorageKey>,
        block: BlockNumber,
    ) -> Result<EIP1186Proof, Self::Error>;
}

/// Data structure with proof for one single storage-entry
#[derive(Debug, Default, Clone, PartialEq, Eq, Deserialize, Serialize)]
pub struct StorageProof {
    pub key: StorageKey,
    pub proof: Vec<Bytes>,
    pub value: StorageValue,
}

impl From<EthersStorageProof> for StorageProof {
    fn from(proof: EthersStorageProof) -> Self {
        StorageProof {
            key: from_ethers_u256(proof.key).to_be_bytes().into(),
            proof: proof.proof.into_iter().map(from_ethers_bytes).collect(),
            value: from_ethers_u256(proof.value),
        }
    }
}

/// Response for EIP-1186 account proof `eth_getProof`
#[derive(Debug, Default, Clone, PartialEq, Eq, Deserialize, Serialize)]
pub struct EIP1186Proof {
    pub address: Address,
    pub balance: U256,
    pub code_hash: B256,
    pub nonce: TxNumber,
    pub storage_hash: B256,
    pub account_proof: Vec<Bytes>,
    pub storage_proof: Vec<StorageProof>,
}
