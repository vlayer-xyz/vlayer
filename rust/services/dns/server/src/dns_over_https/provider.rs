use super::{Query, Response};

pub trait Provider {
    fn resolve(&self, query: &Query) -> Option<Response>;
}
