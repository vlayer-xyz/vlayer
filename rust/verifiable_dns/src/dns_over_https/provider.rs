use super::{Query, Response};

pub trait Provider {
    type Error: Send;

    fn resolve(
        &self,
        query: &Query,
    ) -> impl std::future::Future<Output = Result<Response, Self::Error>> + Send;
}
