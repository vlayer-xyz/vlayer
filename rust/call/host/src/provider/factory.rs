use crate::host::error::HostError;
use crate::provider::{BlockingProvider, EthersProvider};
use alloy_primitives::ChainId;
use std::collections::HashMap;
use std::path::PathBuf;

use super::cache::CachedProvider;
use super::{EthersClient, FileProvider};

pub trait ProviderFactory<P>
where
    P: BlockingProvider,
{
    fn create(&self, chain_id: ChainId) -> Result<P, HostError>;
}

pub struct EthersProviderFactory {
    rpc_urls: HashMap<ChainId, String>,
}

const MAX_RETRY: u32 = 3;
const INITIAL_BACKOFF: u64 = 500;

impl EthersProviderFactory {
    pub fn new(rpc_urls: HashMap<ChainId, String>) -> Self {
        EthersProviderFactory { rpc_urls }
    }
}

impl ProviderFactory<EthersProvider<EthersClient>> for EthersProviderFactory {
    fn create(&self, chain_id: ChainId) -> Result<EthersProvider<EthersClient>, HostError> {
        let url = self
            .rpc_urls
            .get(&chain_id)
            .ok_or(HostError::NoRpcUrl(chain_id))?;

        let client = EthersClient::new_client(url, MAX_RETRY, INITIAL_BACKOFF)?;

        Ok(EthersProvider::new(client))
    }
}

fn get_path(
    rpc_file_cache: &HashMap<ChainId, String>,
    chain_id: ChainId,
) -> Result<PathBuf, HostError> {
    let file_path_str = rpc_file_cache
        .get(&chain_id)
        .ok_or_else(|| HostError::NoRpcCache(chain_id))?;

    Ok(PathBuf::from(file_path_str))
}

pub struct FileProviderFactory {
    rpc_file_cache: HashMap<ChainId, String>,
}

impl FileProviderFactory {
    pub fn new(rpc_file_cache: HashMap<ChainId, String>) -> Self {
        FileProviderFactory { rpc_file_cache }
    }
}

impl ProviderFactory<FileProvider> for FileProviderFactory {
    fn create(&self, chain_id: ChainId) -> Result<FileProvider, HostError> {
        let file_path = get_path(&self.rpc_file_cache, chain_id)?;

        FileProvider::from_file(&file_path).map_err(|err| {
            HostError::Provider(format!(
                "Error creating provider for chain ID {}: {}",
                chain_id, err
            ))
        })
    }
}

pub struct CachedProviderFactory {
    ethers_provider_factory: EthersProviderFactory,
    rpc_file_cache: HashMap<ChainId, String>,
}

impl CachedProviderFactory {
    pub fn new(
        rpc_urls: HashMap<ChainId, String>,
        rpc_file_cache: HashMap<ChainId, String>,
    ) -> Self {
        CachedProviderFactory {
            ethers_provider_factory: EthersProviderFactory::new(rpc_urls),
            rpc_file_cache,
        }
    }
}

impl ProviderFactory<CachedProvider<EthersProvider<EthersClient>>> for CachedProviderFactory {
    fn create(
        &self,
        chain_id: ChainId,
    ) -> Result<CachedProvider<EthersProvider<EthersClient>>, HostError> {
        let file_path = get_path(&self.rpc_file_cache, chain_id)?;

        let provider = self.ethers_provider_factory.create(chain_id)?;
        CachedProvider::new(file_path, provider).map_err(|err| {
            HostError::Provider(format!(
                "Error creating provider for chain ID {}: {}",
                chain_id, err
            ))
        })
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use alloy_chains::Chain;

    #[test]
    fn try_new_invalid_rpc_url() -> anyhow::Result<()> {
        let chain_id = Chain::mainnet().id();
        let rpc_urls = [(chain_id, "http://localhost:123".to_string())]
            .into_iter()
            .collect();
        let factory = EthersProviderFactory::new(rpc_urls);
        let provider = factory.create(chain_id)?;
        let res = provider.get_block_number();
        let error = res.unwrap_err();

        assert!(error.to_string().contains(
            "(http://localhost:123/): error trying to connect: tcp connect error: Connection refused"
        ));

        Ok(())
    }
}
