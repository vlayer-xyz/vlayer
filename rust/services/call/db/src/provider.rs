#![allow(clippy::unwrap_used)]

use std::{
    collections::HashMap,
    sync::{Arc, RwLock},
};

use alloy_primitives::{Address, B256, U256};
use provider::BlockingProvider;
#[allow(clippy::disallowed_types)]
use revm::{
    DatabaseRef,
    primitives::{AccountInfo, Bytecode, KECCAK_EMPTY},
};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum Error {
    #[error("provider error: {0}")]
    Provider(#[from] provider::Error),
    #[error("invalid block number: {0}")]
    InvalidBlockNumber(u64),
}
pub type Result<T> = std::result::Result<T, Error>;

/// A revm [Database] backed by a [Provider].
#[derive(Debug)]
pub struct ProviderDb {
    #[allow(dead_code)]
    pub provider: Arc<dyn BlockingProvider>,
    pub block_number: u64,
    code_hashes: RwLock<HashMap<B256, Address>>,
}

impl ProviderDb {
    /// Creates a new [ProviderDb] with the given provider and block number.
    pub fn new(provider: Arc<dyn BlockingProvider>, block_number: u64) -> Self {
        Self {
            provider,
            block_number,
            code_hashes: RwLock::new(HashMap::new()),
        }
    }
}

impl DatabaseRef for ProviderDb {
    type Error = Error;

    #[allow(clippy::disallowed_types)]
    fn basic_ref(&self, address: Address) -> Result<Option<AccountInfo>> {
        let proof = self
            .provider
            .get_proof(address, vec![], self.block_number)?;
        if proof.code_hash == B256::ZERO {
            return Ok(None);
        }
        #[allow(clippy::expect_used)]
        self.code_hashes
            .write()
            .expect("poisoned lock")
            .insert(proof.code_hash.0.into(), proof.address);
        Ok(Some(AccountInfo {
            nonce: proof.nonce,
            balance: proof.balance,
            code_hash: proof.code_hash,
            code: None,
        }))
    }

    #[allow(clippy::disallowed_types)]
    fn code_by_hash_ref(&self, code_hash: B256) -> Result<Bytecode> {
        if code_hash == KECCAK_EMPTY {
            return Ok(Bytecode::new());
        }
        #[allow(clippy::expect_used)]
        let contract_address = *self
            .code_hashes
            .read()
            .expect("poisoned lock")
            .get(&code_hash)
            .expect("`basic` must be called first for the corresponding account");
        let code = self
            .provider
            .get_code(contract_address, self.block_number)?;
        Ok(Bytecode::new_raw(code.0.into()))
    }

    #[allow(clippy::disallowed_types)]
    fn storage_ref(&self, address: Address, index: U256) -> Result<U256> {
        let storage = self
            .provider
            .get_storage_at(address, index.into(), self.block_number)?;
        Ok(storage)
    }

    #[allow(clippy::disallowed_types)]
    fn block_hash_ref(&self, number: u64) -> Result<B256> {
        let header = self
            .provider
            .get_block_header(number.into())?
            .ok_or(Error::InvalidBlockNumber(number))?;
        Ok(header.hash_slow())
    }
}

pub use self::Error as ProviderDbError;
