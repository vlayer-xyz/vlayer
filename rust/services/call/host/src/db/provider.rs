use std::{
    collections::HashMap,
    fmt::Debug,
    sync::{Arc, RwLock},
};

use alloy_primitives::{Address, B256, U256};
use provider::BlockingProvider;
use revm::{
    primitives::{AccountInfo, Bytecode, KECCAK_EMPTY},
    DatabaseRef,
};
use thiserror::Error;

/// Error type for the [ProviderDb].
#[derive(Error, Debug)]
pub enum ProviderDbError {
    #[error("provider error: {0}")]
    Provider(#[from] provider::Error),
    #[error("invalid block number: {0}")]
    InvalidBlockNumber(u64),
    #[error("hash missing for block: {0}")]
    BlockHashMissing(U256),
}

/// A revm [Database] backed by a [Provider].
pub(crate) struct ProviderDb {
    pub(crate) provider: Arc<dyn BlockingProvider>,
    pub(crate) block_number: u64,

    /// Cache for code hashes to contract addresses.
    code_hashes: RwLock<HashMap<B256, Address>>,
}

impl ProviderDb {
    /// Creates a new [ProviderDb] with the given provider and block number.
    pub(crate) fn new(provider: Arc<dyn BlockingProvider>, block_number: u64) -> Self {
        Self {
            provider,
            block_number,
            code_hashes: RwLock::new(HashMap::new()),
        }
    }
}

impl DatabaseRef for ProviderDb {
    type Error = ProviderDbError;

    fn basic_ref(&self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
        // use `eth_getProof` to get all the account info with a single call
        let proof = self
            .provider
            .get_proof(address, vec![], self.block_number)?;
        // for non-existent accounts, the code hash is zero
        // see https://github.com/ethereum/go-ethereum/issues/28441
        if proof.code_hash == B256::ZERO {
            return Ok(None);
        }
        // cache the code hash to address mapping, so we can later retrieve the code
        self.code_hashes
            .write()
            .expect("poisoned lock")
            .insert(proof.code_hash.0.into(), proof.address);

        Ok(Some(AccountInfo {
            nonce: proof.nonce,
            balance: proof.balance,
            code_hash: proof.code_hash,
            code: None,
        }))
    }

    fn code_by_hash_ref(&self, code_hash: B256) -> Result<Bytecode, Self::Error> {
        // avoid querying the RPC if the code hash is empty
        if code_hash == KECCAK_EMPTY {
            return Ok(Bytecode::new());
        }

        // this works because we always call `basic_ref` first
        let contract_address = *self
            .code_hashes
            .read()
            .expect("poisoned lock")
            .get(&code_hash)
            .expect("`basic` must be called first for the corresponding account");
        let code = self
            .provider
            .get_code(contract_address, self.block_number)?;

        Ok(Bytecode::new_raw(code.0.into()))
    }

    fn storage_ref(&self, address: Address, index: U256) -> Result<U256, Self::Error> {
        let storage = self
            .provider
            .get_storage_at(address, index.into(), self.block_number)?;

        Ok(storage)
    }

    fn block_hash_ref(&self, number: u64) -> Result<B256, Self::Error> {
        let header = self
            .provider
            .get_block_header(number.into())?
            .ok_or(ProviderDbError::InvalidBlockNumber(number))?;

        Ok(header.hash_slow())
    }
}
