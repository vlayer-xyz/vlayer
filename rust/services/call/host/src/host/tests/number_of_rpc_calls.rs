use std::{path::PathBuf, sync::Arc};

use alloy_chains::NamedChain::{Mainnet, OptimismSepolia};
use alloy_primitives::address;
use call_common::ExecutionLocation;
use call_engine::{evm::env::cached::CachedEvmEnv, travel_call};
use call_rpc::rpc_cache_path;
use provider::{CachedMultiProvider, CachedProvider, profiling};

use crate::{
    Call,
    evm_env::factory::HostEvmEnvFactory,
    host::tests::call,
    test_harness::contracts::{
        time_travel::{self, AVERAGE_BALANCE_OF_CALL, SIMPLE_TIME_TRAVEL},
        usdt::{self, IERC20::balanceOfCall, USDT},
    },
};

fn profile(
    chain: &str,
    test_name: &str,
    location: ExecutionLocation,
    call: &Call,
) -> anyhow::Result<profiling::State> {
    let file_path = PathBuf::from(rpc_cache_path(chain, test_name));
    let provider = CachedProvider::from_file(&file_path)?;
    let profiling_provider = Arc::new(profiling::Provider::new(provider));
    let multi_provider =
        CachedMultiProvider::from_provider(location.chain_id, profiling_provider.clone());
    let envs = CachedEvmEnv::from_factory(HostEvmEnvFactory::new(multi_provider));

    let _ = travel_call::Executor::new(&envs, location, true).call(call);

    Ok(profiling_provider.state())
}

#[tokio::test]
async fn usdt_erc20_balance_of() -> anyhow::Result<()> {
    let location: ExecutionLocation = (Mainnet, usdt::BLOCK_NO).into();
    let binance_8 = address!("F977814e90dA44bFA03b6295A0616a897441aceC");
    let call = call(USDT, &balanceOfCall { account: binance_8 });

    let state = profile("mainnet", "usdt_erc20_balance_of", location, &call)?;

    assert_eq!(state.total_count(), 6);
    insta::with_settings!({sort_maps => true}, {
        insta::assert_yaml_snapshot!(state)
    });

    Ok(())
}

#[tokio::test]
async fn time_travel() -> anyhow::Result<()> {
    let location: ExecutionLocation = (OptimismSepolia, time_travel::BLOCK_NO).into();
    let call = call(SIMPLE_TIME_TRAVEL, &AVERAGE_BALANCE_OF_CALL);
    let state = profile("op_sepolia", "time_travel", location, &call)?;

    assert_eq!(state.total_count(), 79);
    insta::with_settings!({sort_maps => true}, {
        insta::assert_yaml_snapshot!(state)
    });

    Ok(())
}
