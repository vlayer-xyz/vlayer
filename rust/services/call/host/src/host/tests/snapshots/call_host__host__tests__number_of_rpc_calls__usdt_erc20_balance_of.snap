---
source: services/call/host/src/host/tests/number_of_rpc_calls.rs
expression: state
---
latest_block: 0
header:
  "0x1297121": 1
balance: {}
code:
  19493153:
    "0xdac17f958d2ee523a2206206994597c13d831ec7": 1
nonce: {}
proof:
  19493153:
    "0x388c818ca8b9251b393131c08a736a67ccb19297": 1
    "0xdac17f958d2ee523a2206206994597c13d831ec7": 1
storage:
  19493153:
    "0xdac17f958d2ee523a2206206994597c13d831ec7":
      "0x000000000000000000000000000000000000000000000000000000000000000a": 1
      "0x0be16d71963429204d70543701f859c43526c316ac005c10114f4694ca405f36": 1
